### PR TITLE
Implement GPUQueue write methods and update wgpu-core

### PR DESCRIPTION
**Pull request by kunalmohan**
Implements `GPUQueue.writeBuffer` and `GPUQueue.writeTexture`.

r?@kvark

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/script/dom/gpuqueue.rs [hunk 1]
```diff
@@ -69,4 +78,105 @@ impl GPUQueueMethods for GPUQueue {
...
+        buffer_offset: GPUSize64,
+        data: CustomAutoRooterGuard<ArrayBuffer>,
+        data_offset: GPUSize64,
+        size: Option<GPUSize64>,
+    ) -> Fallible<()> {
+        let bytes = data.to_vec();
+        let content_size = if let Some(s) = size {
+            s
+        } else {
+            bytes.len() as GPUSize64 - data_offset
```
Comment (kvark) [inline components/script/dom/gpuqueue.rs:96]: this should be taking the buffer size, not the bytes length

File: components/script/dom/gpuqueue.rs [hunk 2]
```diff
@@ -69,4 +78,105 @@ impl GPUQueueMethods for GPUQueue {
...
+        };
+        let valid = data_offset + content_size <= bytes.len() as u64 &&
+            buffer.state() == GPUBufferState::Unmapped &&
+            content_size % 4 == 0 &&
```
Comment (kvark) [inline components/script/dom/gpuqueue.rs:100]: please use the actual constant of `wgc::COPY_BUFFER_ALIGNMENT`

File: components/script/dom/gpuqueue.rs [hunk 3]
```diff
@@ -69,4 +78,105 @@ impl GPUQueueMethods for GPUQueue {
...
+            buffer_offset % 4 == 0 &&
+            buffer.is_valid();
+
+        if !valid {
+            return Err(Error::Operation);
+        }
+
+        let final_data = IpcSharedMemory::from_bytes(
+            &bytes[data_offset as usize..(data_offset + content_size) as usize],
```
Comment (kvark) [inline components/script/dom/gpuqueue.rs:109]: so this should throw an exception if there is not enough data in the array buffer

File: components/script/dom/gpuqueue.rs [hunk 4]
```diff
@@ -69,4 +78,105 @@ impl GPUQueueMethods for GPUQueue {
...
+
+        if !valid {
+            return Err(Error::Operation);
+        }
+
+        let texture_cv = wgpu_com::TextureCopyView {
+            texture: destination.texture.id().0,
+            mip_level: destination.mipLevel,
+            origin: match destination.origin {
+                GPUOrigin3D::RangeEnforcedUnsignedLongSequence(ref v) => wgt::Origin3d {
```
Comment (kvark) [inline components/script/dom/gpuqueue.rs:143]: there needs to be a helper function for this match, I imagine this is not the only place where it's used

**Review by kvark — COMMENTED** [review]
General implementation is all good. There is one spot that needs to be a bit reworked.

File: components/script/dom/gpuqueue.rs (hunk 1 above)
Comment (kunalmohan) [reply components/script/dom/gpuqueue.rs:96]: The spec mentions it to be `data.byteLength` (https://gpuweb.github.io/gpuweb/#dom-gpuqueue-writebuffer). Should that be corrected?

File: components/script/dom/gpuqueue.rs (hunk 3 above)
Comment (kunalmohan) [reply components/script/dom/gpuqueue.rs:109]: A check for this is there on line no. 98
```
let valid = data_offset + content_size <= bytes.len() as u64
```
Or am I missing something?

File: components/script/dom/gpuqueue.rs (hunk 4 above)
Comment (kunalmohan) [reply components/script/dom/gpuqueue.rs:143]: This is the only use at the moment. I was thinking of creating the function once we come across more of its uses.

File: components/script/dom/gpuqueue.rs (hunk 1 above)
Comment (kvark) [reply components/script/dom/gpuqueue.rs:96]: Oh wow, that's a surprise to me :) Thank you for correcting me!

**Review by kvark — APPROVED** [review]
Looks like everything is resolved

File: components/script/dom/gpuqueue.rs (hunk 1 above)
Comment (kvark) [reply components/script/dom/gpuqueue.rs:96]: Filed https://github.com/gpuweb/gpuweb/pull/903 to clarify

Comment (kunalmohan) [reply components/script/dom/gpuqueue.rs:96]: Should I change it to buffer size then?

**Comment by CYBAI** [comment]
@bors-servo retry

- #24622,  #24066, #26895

**Comment by CYBAI** [comment]
@bors-servo retry

- #24066

**Pull request merged**